### PR TITLE
Another step toward full unrolling of loops: cache extracted body

### PR DESCRIPTION
**Pull request by tim-hoffman**

### CODE REVIEW

File: circuit_passes/src/passes/loop_unroll/body_extractor.rs [hunk 1]
```diff
@@ -132,20 +154,37 @@ impl ExtraArgsResult {
...
+    iter_count: usize,
+    extra_arg_info: Rc<ExtraArgsResult>,
+}
+
+#[derive(Debug, Default)]
 pub struct LoopBodyExtractor {
-    new_body_functions: RefCell<Vec<FunctionCode>>,
+    new_body_functions: RefCell<BTreeMap<Key, FunctionCode>>,
+    /// Exists only to stabilize function order for lit test output.
+    func_creation_order: RefCell<Vec<String>>,
```
Comment (tim-hoffman) [inline circuit_passes/src/passes/loop_unroll/body_extractor.rs:168]: If we care about the performance impact of this, we would need to add a cmd line flag for lit testing I think.

File: circuit_passes/src/passes/loop_unroll/body_extractor.rs [hunk 2]
```diff
@@ -74,36 +77,55 @@ impl SubcmpSignalCompare {
...
+    loc_to_itr_to_ref: IndexMap<BucketId, Vec<Option<(AddressType, AddressOffset)>>>,
+    loc_to_args: IndexMap<BucketId, ArgIndex>,
     num_args: usize,
 }
 
+impl PartialOrd for ExtraArgsResult {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        PartialOrd::partial_cmp(&self.num_args, &other.num_args)
+            .and_then(|_| self.loc_to_args.iter().partial_cmp(other.loc_to_args.iter()))
+            .and_then(|_| self.loc_to_itr_to_ref.iter().partial_cmp(other.loc_to_itr_to_ref.iter()))
```
Comment (iangneal) [inline circuit_passes/src/passes/loop_unroll/body_extractor.rs:97]: I could be wrong because I'm not super familiar with this, but I don't think this will end up computing the same ordering as the Ord trait below due to how `and_then` works. `and_then` should chain when the prior `Option` is not `None`, but could be `Some(Less)` or `Some(Greater)`, whereas `then_with` below chains only when the prior `Ordering` returned is `Equal`. So, I think here, you could have the following:
```rust
PartialOrd::partial_cmp(&self.num_args, &other.num_args) // returns Some(Less)
            .and_then(|_| self.loc_to_args.iter().partial_cmp(other.loc_to_args.iter())) // returns Some(Greater)
            .and_then(|_| self.loc_to_itr_to_ref.iter().partial_cmp(other.loc_to_itr_to_ref.iter())) // returns Some(Equal)
```

And your overall result will be `Equal`, but below, you'd have:
```rust
self.num_args
            .cmp(&other.num_args) // Returns Less
            .then_with(|| self.loc_to_args.iter().cmp(other.loc_to_args.iter())) // short-circuited
            .then_with(|| self.loc_to_itr_to_ref.iter().cmp(other.loc_to_itr_to_ref.iter())) // short-circuited
```
And thus returns `Less`.

So maybe we can use the ignored parameters in the `and_then` lambdas so we'll get the same behavior in both cases.

Like I said, could be wrong, but either way, I found the comparison a little hard to follow.

Comment (tim-hoffman) [reply circuit_passes/src/passes/loop_unroll/body_extractor.rs]: @iangneal fixed. Since I had to implement Ord anyway, we can just use it.

File: circuit_passes/src/passes/loop_unroll/body_extractor.rs [hunk 3]
```diff
@@ -596,7 +669,7 @@ impl LoopBodyExtractor {
                 println!("subcmp_arg_groups = {:?}", subcmp_arg_groups);
             }
             //ASSERT: Every bucket mapped to a Some(SubcmpSignal) in any iteration is present in exactly one group.
-            assert!(bucket_to_itr_to_ref
+            assert!(loc_to_itr_to_ref
                 .iter()
                 .filter_map(|(k, v)| {
                     if v.iter().any(|e| matches!(e, Some((AddressType::SubcmpSignal { .. }, _)))) {
```
Comment (0xddom) [inline circuit_passes/src/passes/loop_unroll/body_extractor.rs:675]: Can we make this a debug assert? Looks expensive to iterate over everything

File: circuit_passes/src/passes/loop_unroll/body_extractor.rs [hunk 4]
```diff
@@ -74,115 +84,184 @@ impl SubcmpSignalCompare {
...
+                        "[extract] storing orig loc data for: {}+{:?} -> {:?}",
+                        extracted_name, iter_env, mapping
+                    );
+                }
+                // NOTE: Encountering different iteration counts for the same loop will produce
+                //  different Env at loop header (i.e. `iter_env`) and the same Env at the loop
+                //  header will produce the same reverse mapping. Thus, no key conflicts.
+                checked_insert!(&mut *extraction_data, iter_env, mapping);
+            }
+        }
```
Comment (0xddom) [inline circuit_passes/src/passes/loop_unroll/body_extractor.rs:242]: Can we make this a method of `recorder` instead of making `global_data` public?

**Review by 0xddom — CHANGES_REQUESTED** [review]
Just some code organization thoughts

**Review by iangneal — APPROVED** [review]
LGTM!

File: circuit_passes/src/passes/loop_unroll/body_extractor.rs (hunk 4 above)
Comment (tim-hoffman) [reply circuit_passes/src/passes/loop_unroll/body_extractor.rs:242]: I had done that before but I didn't like it. The function only accesses some data of `recorder` to store to the global map so I thought it was cleaner to move it out of `recorder` and just use the accessor functions. If you don't like the public field I could just make an accessor function for `recorder.global_data` but I thought this was cleaner than just adding a boilerplate function since the field is not mutable anyway so it cannot be accidentally misused by a future code change.

File: circuit_passes/src/passes/loop_unroll/body_extractor.rs (hunk 1 above)
Comment (0xddom) [reply circuit_passes/src/passes/loop_unroll/body_extractor.rs:169]: We could go back to a manual CI script then, to get more control. I honestly don't like leaving nix handle everything

File: circuit_passes/src/passes/loop_unroll/body_extractor.rs (hunk 4 above)
Comment (0xddom) [reply circuit_passes/src/passes/loop_unroll/body_extractor.rs:242]: I don't like getters for the sake of getters, so we can leave it as is then

**Review by 0xddom — APPROVED** [review]
We gucci
